pkg/controller/transaction/controller.go

```diff
@@ -0,0 +1,228 @@
+func NewController(topo topo.Store, transactions transaction.Store, configurations configuration.Store, pluginRegistry *pluginregistry.PluginRegistry) *controller.Controller {
+	c := controller.NewController("transaction")
+
+	c.Watch(&Watcher{
+		transactions: transactions,
+	})
+
+	c.Watch(&ConfigurationWatcher{
+		transactions:   transactions,
+		configurations: configurations,
+	})
+
+	c.Reconcile(&Reconciler{
+		transactions:   transactions,
+		topo:           topo,
+		configurations: configurations,
+		pluginRegistry: pluginRegistry,
+	})
+
+	return c
+}
+
+// Reconciler reconciles transactions
+type Reconciler struct {
+	topo           topo.Store
+	transactions   transaction.Store
+	configurations configuration.Store
+	pluginRegistry *pluginregistry.PluginRegistry
+}
+
+// Reconcile reconciles transactions
+func (r *Reconciler) Reconcile(id controller.ID) (controller.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	transactionID := id.Value.(configapi.TransactionID)
+	transaction, err := r.transactions.Get(ctx, transactionID)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Warnf("Failed to reconcile transaction %s, %s", transactionID, err)
+			return controller.Result{}, err
+		}
+		log.Debugf("Transaction %s not found", transactionID)
+		return controller.Result{}, nil
+	}
+
+	log.Infof("Reconciling transaction %v", transaction)
+	if ok, err := r.reconcileTransaction(ctx, transaction); err != nil {
+		log.Warnf("Failed to reconcile transaction: %s, %s", transactionID, err)
+		return controller.Result{}, err
+	} else if ok {
+		return controller.Result{}, nil
+	}
+
+	return controller.Result{}, nil
+}
+
+func (r *Reconciler) reconcileTransaction(ctx context.Context, transaction *configapi.Transaction) (bool, error) {
+
+	log.Debugf("Reconciling transaction %s in %s state", transaction.ID, transaction.Status.State)
+	// If the transaction is Pending, begin validation if the prior transaction
+	// has already been applied. This simplifies concurrency control in the controller
+	// and guarantees transactions are applied to the configurations in sequential order.
+	if transaction.Status.State == configapi.TransactionState_TRANSACTION_PENDING {
+		if transaction.Index > 1 {
+			prevTransaction, err := r.transactions.GetByIndex(ctx, transaction.Index-1)
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					return false, err
+				}
+				return false, nil
+			}
+			if prevTransaction.Status.State == configapi.TransactionState_TRANSACTION_COMPLETE ||
+				prevTransaction.Status.State == configapi.TransactionState_TRANSACTION_FAILED {
+				transaction.Status.State = configapi.TransactionState_TRANSACTION_VALIDATING
+				err = r.transactions.Update(ctx, transaction)
+				if err != nil {
+					if !errors.IsConflict(err) || !errors.IsNotFound(err) {
+						return false, err
+					}
+					return false, nil
+				}
+
+				return true, nil
+			}
+		}
```

Comment:
There’s something missing here: this covers if the transaction index is > 1, but what if it’s 1? Then we can proceed to validate and apply it.

But also we probably shouldn’t use the index 1, but rather consider this the first transaction in the log if there’s no prior transaction. Once we introduce log compaction, the first index in the log will not necessarily be `1`.